dbbackup/dbbackup.py

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env python3
+from lightning import Plugin
+import os
+import shutil
+import sqlite3
+from stat import S_IRUSR, S_IWUSR
+
+
+plugin = Plugin()
+plugin.sqlite_pre_init_cmds = []
+plugin.initted = False
+
+
+class NewBackupFileError(Exception):
+    def __init__(self, path, e):
+        self.message = 'Could not create db-backup-file {} : {}'.format(path, e)
+
+
+# Create fresh backup, started as a copy
+def new_backup_file(db, backup):
+    try:
+        shutil.copyfile(db, backup)
+        os.chmod(backup, S_IRUSR | S_IWUSR)     # rw permission for owner
+        plugin.log('Creating new db-backup-file: {}'.format(backup))
+    except Exception as e:
+        raise NewBackupFileError(backup, e)
+
+
+@plugin.init()
+def init(configuration, options, plugin):
+    # FIXME: `==` should be changed to `is not None`, see workaround below
+    if plugin.get_option('db-backup-file') == '':
+        plugin.log('No db-backup-file specified', 'error')
+        plugin.rpc.stop()   # stop lightningd
+
+    try:
+        db = os.path.join(configuration['lightning-dir'], 'lightningd.sqlite3')
+        backup = plugin.get_option('db-backup-file')
+
+        # If backup exist, replay pre_init_cmds on a temporary copy
+        if os.path.isfile(backup):
+            plugin.log('Found existing db-backup-file: {} comparing...'.format(backup))
+            backup_copy = shutil.copy(backup, backup + '.tmp')
+            db1 = sqlite3.connect(backup_copy, isolation_level=None)
+            db2 = sqlite3.connect('file:{}?mode=ro'.format(db), uri=True)   # open in read-only
+            for c in plugin.sqlite_pre_init_cmds:
+                db1.execute(c)
+
+            # If it then matches orignal db, replace backup with copy ... else abort
+            dbs_match = [x for x in db1.iterdump()] == [x for x in db2.iterdump()]
+            db1.close()
+            db2.close()
+            if dbs_match:
+                os.rename(backup_copy, backup)
+                plugin.log("Existing db-backup-file OK and successfully synced")
+            else:
+                plugin.log("Existing db-backup-file differs from original database, i.e. applying"
+                           " pre-init statements (to a copy) didn't make it match the original db", 'error')
+                os.remove(backup_copy)
+                plugin.rpc.stop()   # stop lightningd
+
+        else:
+            new_backup_file(db, backup)
+
+        plugin.conn = sqlite3.connect(backup, isolation_level=None)
+        plugin.initted = True
+        plugin.log('Initialized')
+    except Exception as e:
+        if isinstance(e, NewBackupFileError):
+            plugin.log(e.message, 'error')
+        else:
+            plugin.log('Initialization failed: {}'.format(e), 'error')
+
+        plugin.rpc.stop()   # stop lightningd
+
+
+@plugin.hook('db_write')
+def db_write(plugin, writes):
+    if not plugin.initted:
+        plugin.sqlite_pre_init_cmds += writes
```

Comment:
There is a potential (but unlikely) race condition where `db_write` is invoked before `init`, then C-lightning crashes before it is able to `init` the plugins.  It may be safer to save these in a temporary file that you clear out once `init` is invoked.

Reply:
Good catch, though writes before the `init` call will all be migrations iirc, so these are not catastrophic.

Reply:
Not keeping up with C-lightning upgrades is "not catastrophic"?  Upgrades getting lost in the stored database are arguably worse.  Further, upgrades are a risk by themselves as they represent changed code, and the user might be running on some obscure combination of system components that happen to work on older releases of C-lightning but not on newer ones, so a new release (which is when we are likely to have migrations running) is at increased risk of crashes before `init`... *shrug*

Reply:
@ZmnSCPxj Thanks, there is also a potential dead-lock when the plugin makes the `rpc.stop()` call from `init` while at the same time `lightningd` is waiting for a `db_write` hook to return. TBH I was totally unaware that plugin with (sync) hooks shouldn't make **any** (sync) RPC calls, at least not from the same threat that also handles the hook.

If I understand correctly, RPC calls can be send after [jsonrpc_listen](https://github.com/ElementsProject/lightning/blob/804aee766e968d27fa117a3b21dca6269e78aa7d/lightningd/lightningd.c#L777) but are not be answered until the main [io_loop_with_timers](https://github.com/ElementsProject/lightning/blob/804aee766e968d27fa117a3b21dca6269e78aa7d/lightningd/lightningd.c#L834)

To solve these and similar issues, I am exploring a _sync init_ in [plugins_config](https://github.com/ElementsProject/lightning/blob/804aee766e968d27fa117a3b21dca6269e78aa7d/lightningd/plugin.c#L1066) where `lightningd` waits for the `init` response of certain (perhaps all static?) plugins before continue startup, in a similar way as it does for `getmanifest`.  And maybe shutdown CL when a plugin's `init` returns False? Will make a PR if that doesn't break too much.